File: user_monitor.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <signal.h>
#include <errno.h>
#include <string.h>
#include <time.h>

#include <bpf/libbpf.h>
#include <bpf/bpf.h>

#include "syscall_monitoring.skel.h"

/* libcap runtime checks removed to simplify runtime behavior. */

// Use shared syscall id header to avoid duplication/drift
#include "syscall_ids.h"

// Map older SCID_* names used in this file to the shared header names
#define SCID_SENDMSG    SC_SENDMSG
#define SCID_SENDTO     SC_SENDTO
#define SCID_RECVMSG    SC_RECVMSG
#define SCID_RECVFROM   SC_RECVFROM
#define SCID_CONNECT    SC_CONNECT
#define SCID_CLOSE      SC_CLOSE
#define SCID_CLOSE_RANGE SC_CLOSE_RANGE
#define SCID_READ       SC_READ
#define SCID_READV      SC_READV
#define SCID_WRITE      SC_WRITE
#define SCID_WRITEV     SC_WRITEV
#define SCID_OPEN       SC_OPEN
#define SCID_OPENAT     SC_OPENAT
#define SCID_FSTAT      SC_FSTAT
#define SCID_FSTATAT    SC_FSTATAT
#define SCID_POLL       SC_POLL
#define SCID_PPOLL      SC_PPOLL
#define SCID_EPOLL_WAIT SC_EPOLL_WAIT

// Mirror of kernel structs
struct syscall_stats {
    __u64 count;
    __u64 sum_ns;
    __u64 max_ns;
    __u64 bytes;
};

struct syscall_key {
    __u32 pid;
    __u32 id;
};

static volatile sig_atomic_t exiting = 0;
static void sig_handler(int sig) { exiting = 1; }

// Minimal syscall name table for common syscalls. If not found, print numeric id.
static const char *syscall_name(__u32 id) {
    switch (id) {
    case SCID_SENDMSG:    return "sendmsg";
    case SCID_SENDTO:     return "sendto";
    case SCID_RECVMSG:    return "recvmsg";
    case SCID_RECVFROM:   return "recvfrom";
    case SCID_CONNECT:    return "connect";
    case SCID_CLOSE:      return "close";
    case SCID_CLOSE_RANGE:return "close_range";
    case SCID_READ:       return "read";
    case SCID_READV:      return "readv";
    case SCID_WRITE:      return "write";
    case SCID_WRITEV:     return "writev";
    case SCID_OPEN:       return "open";
    case SCID_OPENAT:     return "openat";
    case SCID_FSTAT:      return "fstat";
    case SCID_FSTATAT:    return "fstatat";
    case SCID_POLL:       return "poll";
    case SCID_PPOLL:      return "ppoll";
    case SCID_EPOLL_WAIT: return "epoll_wait";
    default: return NULL;
    }
}

// Helper: return non-zero if this syscall id has a meaningful "bytes" value
static int syscall_has_bytes(__u32 id) {
    switch (id) {
    case SCID_SENDMSG:
    case SCID_SENDTO:
    case SCID_RECVMSG:
    case SCID_RECVFROM:
    case SCID_READ:
    case SCID_READV:
    case SCID_WRITE:
    case SCID_WRITEV:
        return 1;
    default:
        return 0;
    }
}


int print_and_clear_map(int map_fd) {
    struct syscall_key prev_key;
    struct syscall_key key;
    struct syscall_key next_key;
    struct syscall_stats stats;

    // Iterate map: start with NULL previous key to get the first entry
    // then repeatedly get the next key using the last seen key.
    // This avoids printing multiple times per PID/Call.
    int res = bpf_map_get_next_key(map_fd, NULL, &key);
    while (res == 0) {
        if (bpf_map_lookup_elem(map_fd, &key, &stats) != 0) {
            // couldn't read this entry, try to advance to next
            res = bpf_map_get_next_key(map_fd, &key, &key);
            continue;
        }

        double avg_ns = 0.0;
        if (stats.count)
            avg_ns = (double)stats.sum_ns / (double)stats.count;

        const char *name = syscall_name(key.id);
        char name_buf[32];
        if (!name) {
            snprintf(name_buf, sizeof(name_buf), "sys_%u", key.id);
            name = name_buf;
        }

        /* Read /proc/<pid>/comm to get the task's command name (TASK_COMM_LEN=16) */
        char comm[16] = "-";
        char comm_path[64];
        snprintf(comm_path, sizeof(comm_path), "/proc/%u/comm", key.pid);
        FILE *f = fopen(comm_path, "r");
        if (f) {
            if (fgets(comm, sizeof(comm), f) != NULL) {
                /* strip trailing newline if present */
                size_t l = strlen(comm);
                if (l && comm[l-1] == '\n') comm[l-1] = '\0';
            }
            fclose(f);
        }

     if (syscall_has_bytes(key.id)) {
         printf("PID=%u comm=%s call=%s count=%llu avg_ms=%.3f max_ms=%.3f bytes=%llu\n",
             key.pid,
             comm,
             name,
             (unsigned long long)stats.count,
             avg_ns / 1e6,
             (double)stats.max_ns / 1e6,
             (unsigned long long)stats.bytes);
     } else {
         printf("PID=%u comm=%s call=%s count=%llu avg_ms=%.3f max_ms=%.3f\n",
             key.pid,
             comm,
             name,
             (unsigned long long)stats.count,
             avg_ns / 1e6,
             (double)stats.max_ns / 1e6);
     }

        // delete entry so stats are flushed
        if (bpf_map_delete_elem(map_fd, &key) != 0) {
            // couldn't delete, nothing more we can do for this entry
        }

        // advance to next key
        prev_key = key;
        res = bpf_map_get_next_key(map_fd, &prev_key, &next_key);
        if (res == 0)
            key = next_key;
    }

    return 0;
}

// Runtime libcap checks removed; permission checks remain when loading/attaching BPF.

int main(int argc, char **argv) {
    struct syscall_monitoring_bpf *skel;
    int err;
    unsigned int interval = 10; /* seconds, default */
    // Simple dynamic array of comm filters (max 64 per BPF map definition)
    const char *comm_filters[64];
    int comm_filter_count = 0;

    /* parse user arguments: --interval / -i <seconds> */
    for (int i = 1; i < argc; i++) {
        if ((strcmp(argv[i], "-i") == 0 || strcmp(argv[i], "--interval") == 0) && i + 1 < argc) {
            char *endptr = NULL;
            long val = strtol(argv[i+1], &endptr, 10);
            if (endptr == argv[i+1] || val <= 0) {
                fprintf(stderr, "invalid interval '%s'\n", argv[i+1]);
                return 1;
            }
            interval = (unsigned int)val;
            i++; /* skip value */
            continue;
        }
        if ((strcmp(argv[i], "-c") == 0 || strcmp(argv[i], "--comm") == 0) && i + 1 < argc) {
            if (comm_filter_count >= 64) {
                fprintf(stderr, "too many -c/--comm filters (max 64)\n");
                return 1;
            }
            comm_filters[comm_filter_count++] = argv[++i];
            continue;
        }
        if (strcmp(argv[i], "-h") == 0 || strcmp(argv[i], "--help") == 0) {
            printf("usage: %s [-i|--interval seconds] [-c|--comm name]...\n", argv[0]);
            printf("  -i, --interval <sec>  Flush/aggregation interval (default %u)\n", interval);
            printf("  -c, --comm <name>     Only monitor processes whose task comm matches <name>.\n");
            printf("                       May be supplied multiple times (up to 64). If omitted, monitors all.\n");
            return 0;
        }
    }

    signal(SIGINT, sig_handler);
    signal(SIGTERM, sig_handler);

    /* Note: runtime libcap checks were removed. Permissions errors while
     * loading or attaching BPF programs will still be reported below. */

    skel = syscall_monitoring_bpf__open();
    if (!skel) {
        fprintf(stderr, "failed to open BPF skeleton\n");
        return 1;
    }

    err = syscall_monitoring_bpf__load(skel);
    if (err) {
        /* libbpf often returns -1 and prints a helpful message to stderr
         * about the underlying cause; still, provide actionable hints to
         * the user for common causes: missing capabilities / not running
         * as root, or RLIMIT_MEMLOCK too low. */
        fprintf(stderr, "failed to load BPF object: %d\n", err);
        if (errno == EPERM || errno == EACCES) {
            fprintf(stderr, "       Permission denied while loading BPF object.\n");
            fprintf(stderr, "       Try running as root: sudo %s\n", argv[0]);
            fprintf(stderr, "       Or grant capabilities: sudo setcap cap_bpf,cap_perfmon,cap_sys_resource+ep %s\n", argv[0]);
        } else if (errno == ENOMEM) {
            fprintf(stderr, "       Insufficient memory or RLIMIT_MEMLOCK too low. Increase with:\n");
            fprintf(stderr, "         sudo prlimit --pid $$ --memlock=unlimited\n");
            fprintf(stderr, "       Or add to /etc/security/limits.conf: \"* hard memlock unlimited\"\n");
        } else {
            fprintf(stderr, "       See the libbpf output above for details; ensure your kernel supports BPF and that RLIMIT_MEMLOCK is large enough.\n");
        }
        goto cleanup;
    }

    err = syscall_monitoring_bpf__attach(skel);
    if (err) {
        if (err == -EACCES || err == -EPERM) {
            fprintf(stderr, "error: failed to attach BPF programs: permission denied (EACCES/EPERM)\n");
            fprintf(stderr, "       This usually means the process lacks CAP_BPF/CAP_PERFMON or isn't running as root.\n");
            fprintf(stderr, "       You can either run as root (sudo %s) or grant capabilities:\n", argv[0]);
            fprintf(stderr, "         sudo setcap cap_bpf,cap_perfmon,cap_sys_resource+ep %s\n", argv[0]);
            fprintf(stderr, "       On some distros you may also need to enable tracing/proc fs or kernel config for BPF.\n");
        } else {
            fprintf(stderr, "failed to attach BPF programs: %d\n", err);
        }
        goto cleanup;
    }

    int map_fd = bpf_map__fd(skel->maps.syscall_stats_map);
    if (map_fd < 0) {
        fprintf(stderr, "failed to get map fd\n");
        goto cleanup;
    }

    /* Set our own PID into the BPF monitor_pid_map so the BPF program
     * ignores this process (prevents the monitor from monitoring itself).
     */
    int mon_map_fd = bpf_map__fd(skel->maps.monitor_pid_map);
    if (mon_map_fd >= 0) {
        __u32 key = 0;
        __u32 val = (__u32)getpid();
        if (bpf_map_update_elem(mon_map_fd, &key, &val, BPF_ANY) != 0) {
            fprintf(stderr, "warning: failed to populate monitor_pid_map: %s\n", strerror(errno));
            /* Not fatal: continue without self-filtering */
        }
        else {
            /* read back for verification */
            __u32 readback = 0;
            if (bpf_map_lookup_elem(mon_map_fd, &key, &readback) == 0) {
                fprintf(stderr, "monitor_pid_map populated with PID=%u\n", readback);
            } else {
                fprintf(stderr, "warning: failed to read back monitor_pid_map: %s\n", strerror(errno));
            }
        }
    } else {
        /* Not fatal: continue without self-filtering */
    }

    /* Remove any existing stats for the monitor PID so we don't print
     * events that were recorded before we populated the monitor map. */
    if (mon_map_fd >= 0 && map_fd >= 0) {
        __u32 key0 = 0;
        __u32 readback = 0;
        if (bpf_map_lookup_elem(mon_map_fd, &key0, &readback) == 0) {
            struct syscall_key skey;
            struct syscall_key sknext;
            int r = bpf_map_get_next_key(map_fd, NULL, &skey);
            while (r == 0) {
                if (skey.pid == readback) {
                    bpf_map_delete_elem(map_fd, &skey);
                }
                r = bpf_map_get_next_key(map_fd, &skey, &sknext);
                if (r == 0)
                    skey = sknext;
            }
        }
    }

    // Populate allowed comms map if filters provided
    if (comm_filter_count > 0) {
        int comm_map_fd = bpf_map__fd(skel->maps.allowed_comms_map);
        int filter_flag_fd = bpf_map__fd(skel->maps.filter_enabled_map);
        if (comm_map_fd < 0 || filter_flag_fd < 0) {
            fprintf(stderr, "error: BPF object missing required maps for comm filtering\n");
            goto cleanup;
        }
        __u32 zero = 0; __u8 one = 1;
        if (bpf_map_update_elem(filter_flag_fd, &zero, &one, BPF_ANY) != 0) {
            fprintf(stderr, "warning: failed to enable comm filtering: %s\n", strerror(errno));
        }
        for (int j = 0; j < comm_filter_count; j++) {
            char key[16] = {0};
            size_t len = strlen(comm_filters[j]);
            if (len >= sizeof(key)) len = sizeof(key) - 1; // truncate silently to match task comm behavior
            memcpy(key, comm_filters[j], len);
            __u8 val = 1;
            if (bpf_map_update_elem(comm_map_fd, &key, &val, BPF_ANY) != 0) {
                fprintf(stderr, "warning: failed adding comm filter '%s': %s\n", comm_filters[j], strerror(errno));
            }
        }
    }

    printf("syscall monitor started, flushing every %us%s. Ctrl-C to exit.\n",
           interval,
           comm_filter_count ? " (filtered by comm)" : "");

    while (!exiting) {
        for (unsigned int i = 0; i < interval && !exiting; i++)
            sleep(1);

        if (exiting) break;

        time_t t = time(NULL);
        printf("\n=== stats flush @ %s", ctime(&t));
        print_and_clear_map(map_fd);
        fflush(stdout);
    }

cleanup:
    syscall_monitoring_bpf__destroy(skel);
    return err != 0;
}

File: user_monitor.go
package main

import (
    "bufio"
    "encoding/binary"
    "errors"
    "flag"
    "fmt"
    "log"
    "os"
    "os/signal"
    "path/filepath"
    "strings"
    "syscall"
    "time"

    "github.com/cilium/ebpf"
    "github.com/cilium/ebpf/link"
    "github.com/cilium/ebpf/rlimit"
)

// Keep in sync with BPF-side structs and IDs.
type syscallStats struct {
    Count uint64
    SumNs uint64
    MaxNs uint64
    Bytes uint64
}

type syscallKey struct {
    PID uint32
    ID  uint32
}

const (
    scSendmsg   = 1
    scSendto    = 2
    scRecvmsg   = 3
    scRecvfrom  = 4
    scConnect   = 5
    scClose     = 6
    scCloseRng  = 7
    scRead      = 8
    scReadv     = 9
    scWrite     = 10
    scWritev    = 11
    scOpen      = 12
    scOpenat    = 13
    scFstat     = 14
    scFstatat   = 15
    scPoll      = 16
    scPpoll     = 17
    scEpollWait = 18
)

func scName(id uint32) string {
    switch id {
    case scSendmsg:
        return "sendmsg"
    case scSendto:
        return "sendto"
    case scRecvmsg:
        return "recvmsg"
    case scRecvfrom:
        return "recvfrom"
    case scConnect:
        return "connect"
    case scClose:
        return "close"
    case scCloseRng:
        return "close_range"
    case scRead:
        return "read"
    case scReadv:
        return "readv"
    case scWrite:
        return "write"
    case scWritev:
        return "writev"
    case scOpen:
        return "open"
    case scOpenat:
        return "openat"
    case scFstat:
        return "fstat"
    case scFstatat:
        return "fstatat"
    case scPoll:
        return "poll"
    case scPpoll:
        return "ppoll"
    case scEpollWait:
        return "epoll_wait"
    default:
        return fmt.Sprintf("sys_%d", id)
    }
}

func scHasBytes(id uint32) bool {
    switch id {
    case scSendmsg, scSendto, scRecvmsg, scRecvfrom, scRead, scReadv, scWrite, scWritev:
        return true
    default:
        return false
    }
}

// Utilities for interacting with BPF maps without generated bindings.
func mustMap(obj *ebpf.CollectionSpec, name string) *ebpf.MapSpec {
    if ms, ok := obj.Maps[name]; ok {
        return ms
    }
    log.Fatalf("BPF object missing map %q", name)
    return nil
}

func main() {
    // Flags compatible with the C version.
    interval := flag.Uint("interval", 10, "Flush interval in seconds")
    flag.UintVar(interval, "i", 10, "Flush interval in seconds")
    var commFilters multiFlag
    flag.Var(&commFilters, "comm", "Only monitor processes with this comm (repeatable)")
    flag.Var(&commFilters, "c", "Only monitor processes with this comm (repeatable)")
    flag.Parse()

    // Bump RLIMIT_MEMLOCK for older kernels
    _ = rlimit.RemoveMemlock()

    // Load the pre-compiled BPF object file.
    objPath := filepath.Join(".", "syscall_monitoring.bpf.o")
    spec, err := ebpf.LoadCollectionSpec(objPath)
    if err != nil {
        log.Fatalf("failed to load BPF collection spec from %s: %v", objPath, err)
    }

    // Create the collection (maps and programs)
    coll, err := ebpf.NewCollection(spec)
    if err != nil {
        // Common EPERM hints similar to C tool
        if errors.Is(err, syscall.EPERM) || errors.Is(err, syscall.EACCES) {
            log.Fatalf("permission denied creating BPF collection: %v\nTry: sudo setcap cap_bpf,cap_perfmon,cap_sys_resource+ep <this_binary> or run as root.", err)
        }
        log.Fatalf("failed to create BPF collection: %v", err)
    }
    defer coll.Close()

    // Attach tracepoint programs. Iterate the loaded collection's programs
    // (coll.Programs) and derive the tracepoint event name from the
    // program name. Use an exceptions map for cases where the tracepoint
    // name differs from the simple convention (e.g. fstat -> newfstat).
    var links []link.Link
    exceptions := map[string]string{
        "enter_fstat":  "sys_enter_newfstat",
        "exit_fstat":   "sys_exit_newfstat",
        "enter_fstatat": "sys_enter_newfstatat",
        "exit_fstatat":  "sys_exit_newfstatat",
    }

    for name, prog := range coll.Programs {
        var ev string
        if e, ok := exceptions[name]; ok {
            ev = e
        } else if strings.HasPrefix(name, "enter_") {
            ev = "sys_enter_" + strings.TrimPrefix(name, "enter_")
        } else if strings.HasPrefix(name, "exit_") {
            ev = "sys_exit_" + strings.TrimPrefix(name, "exit_")
        } else {
            // Not a tracepoint program we expect to attach
            continue
        }

        cat := "syscalls"
        evPath := filepath.Join("/sys/kernel/tracing/events", cat, ev, "id")
        if _, err := os.Stat(evPath); err != nil {
            log.Printf("skipping attach syscalls/%s via program %s: tracepoint not present (%v)", ev, name, err)
            continue
        }

        l, err := link.Tracepoint(cat, ev, prog, nil)
        if err != nil {
            log.Printf("attach syscalls/%s via program %s failed: %v", ev, name, err)
            continue
        }
        links = append(links, l)
    }
    defer func() {
        for _, l := range links {
            _ = l.Close()
        }
    }()

    // Convenience handles for maps we need to touch directly
    statsMap := coll.Maps["syscall_stats_map"]
    monMap := coll.Maps["monitor_pid_map"]
    allowedComms := coll.Maps["allowed_comms_map"]
    filterEnabled := coll.Maps["filter_enabled_map"]
    if statsMap == nil || monMap == nil || allowedComms == nil || filterEnabled == nil {
        // allowed_comms_map/filter_enabled_map may be needed only if filtering used, but assert presence like C.
        log.Fatalf("BPF maps missing: stats:%v mon:%v comms:%v filter:%v", statsMap != nil, monMap != nil, allowedComms != nil, filterEnabled != nil)
    }

    // Set monitor PID to avoid self-monitoring (ouroborus effect)
    key32 := uint32(0)
    self := uint32(os.Getpid())
    if err := monMap.Update(&key32, &self, ebpf.UpdateAny); err != nil {
        log.Printf("warning: failed to populate monitor_pid_map: %v", err)
    } else {
        var rb uint32
        if err := monMap.Lookup(&key32, &rb); err == nil {
            log.Printf("monitor_pid_map populated with PID=%d", rb)
        }
    }

    // Clear any existing stats that already recorded our PID
    if err := clearStatsForPID(statsMap, self); err != nil {
        log.Printf("warning: failed to clear initial stats for self: %v", err)
    }

    // If comm filters provided, enable filtering and populate map
    if len(commFilters) > 0 {
        one := uint8(1)
        if err := filterEnabled.Update(&key32, &one, ebpf.UpdateAny); err != nil {
            log.Printf("warning: enabling comm filter failed: %v", err)
        }
        for _, c := range commFilters {
            k := fixed16(c)
            v := uint8(1)
            if err := allowedComms.Update(&k, &v, ebpf.UpdateAny); err != nil {
                log.Printf("warning: adding comm filter %q failed: %v", c, err)
            }
        }
    }

    log.Printf("syscall monitor started, flushing every %ds%s. Ctrl-C to exit.", *interval, func() string {
        if len(commFilters) > 0 { return " (filtered by comm)" }
        return ""
    }())

    // Ticker loop and signal handling
    sigCh := make(chan os.Signal, 2)
    signal.Notify(sigCh, syscall.SIGINT, syscall.SIGTERM)

    ticker := time.NewTicker(time.Duration(*interval) * time.Second)
    defer ticker.Stop()

    for {
        select {
        case <-ticker.C:
            fmt.Printf("\n=== stats flush @ %s", time.Now().Format(time.RFC1123))
            if err := dumpAndClearStats(statsMap); err != nil {
                log.Printf("flush error: %v", err)
            }
        case <-sigCh:
            return
        }
    }
}

// Iterate a hash map by repeatedly calling NextKey and Lookup, printing and then deleting each entry.
func dumpAndClearStats(m *ebpf.Map) error {
    it := m.Iterate()
    var key syscallKey
    var val syscallStats
    var entries []struct{
        k syscallKey
        v syscallStats
    }
    for it.Next(&key, &val) {
        // copy key/val because iterator reuses the memory
        entries = append(entries, struct{ k syscallKey; v syscallStats }{k: key, v: val})
    }
    if err := it.Err(); err != nil {
        return err
    }

    // Now print and delete outside of the iterator to avoid aborting it
    for _, e := range entries {
        comm := readComm(int(e.k.PID))
        avgMs := 0.0
        if e.v.Count > 0 {
            avgMs = float64(e.v.SumNs) / float64(e.v.Count) / 1e6
        }
        maxMs := float64(e.v.MaxNs) / 1e6
        name := scName(e.k.ID)

        if scHasBytes(e.k.ID) {
            fmt.Printf("PID=%d comm=%s call=%s count=%d avg_ms=%.3f max_ms=%.3f bytes=%d\n",
                e.k.PID, comm, name, e.v.Count, avgMs, maxMs, e.v.Bytes)
        } else {
            fmt.Printf("PID=%d comm=%s call=%s count=%d avg_ms=%.3f max_ms=%.3f\n",
                e.k.PID, comm, name, e.v.Count, avgMs, maxMs)
        }

        kcopy := e.k
        if err := m.Delete(&kcopy); err != nil {
            // Not fatal, continue
        }
    }
    return nil
}

func clearStatsForPID(m *ebpf.Map, pid uint32) error {
    it := m.Iterate()
    var key syscallKey
    var val syscallStats
    var keysToDelete []syscallKey
    for it.Next(&key, &val) {
        if key.PID == pid {
            keysToDelete = append(keysToDelete, key)
        }
    }
    if err := it.Err(); err != nil {
        return err
    }
    for _, k := range keysToDelete {
        kk := k
        _ = m.Delete(&kk)
    }
    return nil
}

// fixed16 returns a 16-byte array filled with the string bytes, truncated or zero-padded.
func fixed16(s string) [16]byte {
    var b [16]byte
    copy(b[:], []byte(s))
    return b
}

// readComm reads /proc/<pid>/comm up to 16 chars similar to C version.
func readComm(pid int) string {
    f, err := os.Open(filepath.Join("/proc", fmt.Sprint(pid), "comm"))
    if err != nil {
        return "-"
    }
    defer f.Close()
    rd := bufio.NewReader(f)
    line, _ := rd.ReadString('\n')
    line = strings.TrimSpace(line)
    if len(line) > 16 {
        line = line[:16]
    }
    return line
}

// multiFlag supports repeated -c/--comm options.
type multiFlag []string

func (m *multiFlag) String() string { return strings.Join(*m, ",") }
func (m *multiFlag) Set(v string) error {
    *m = append(*m, v)
    return nil
}

// Ensure struct layout matches C. Add a sanity check using binary size.
func init() {
    if binary.Size(syscallStats{}) != 8*4 {
        // Four uint64 fields
        panic("syscallStats size mismatch; check field types and packing")
    }
    if binary.Size(syscallKey{}) != 8 {
        // two uint32
        panic("syscallKey size mismatch; check field types and packing")
    }
}
